addons/mrp/models/mrp_production.py

```diff
@@ -1232,66 +1263,84 @@ def _plan_workorders(self, replan=False):
+                vals = {}
+                for workcenter in workcenters:
+                    # Compute theoretical duration
+                    if workorder.workcenter_id == workcenter:
+                        duration_expected = workorder.duration_expected
+                    else:
+                        duration_expected = workorder._get_duration_expected(alternative_workcenter=workcenter)
+                    from_date, to_date = workcenter._get_first_available_slot(start_date, duration_expected)
+                    # If the workcenter is unavailable, try planning on the next one
+                    if not from_date:
+                        continue
+                    # Check if this workcenter is better than the previous ones
+                    if to_date and to_date < workorder_best_finished_date:
+                        workorder_best_start_date = from_date
+                        workorder_best_finished_date = to_date
+                        workorder_best_workcenter = workcenter
+                        vals = {
+                            'workcenter_id': workcenter.id,
+                            'duration_expected': duration_expected,
+                        }
+                # If none of the workcenter are available, raise
+                if workorder_best_finished_date == datetime.datetime.max:
+                    raise UserError(_('Impossible to plan the workorder. Please check the workcenter availabilities.'))
+                # Create leave on chosen workcenter calendar
+                leave = self.env['resource.calendar.leaves'].create({
+                    'name': workorder.display_name,
+                    'calendar_id': workorder_best_workcenter.resource_calendar_id.id,
+                    'date_from': workorder_best_start_date,
+                    'date_to': workorder_best_finished_date,
+                    'resource_id': workorder_best_workcenter.resource_id.id,
+                    'time_type': 'other'
+                })
+                vals['leave_id'] = leave.id
+                workorder.write(vals)
+                # Check if this workorder terminates later than the others in this step
+                if workorder_best_finished_date > step_worst_finished_date:
+                    step_worst_finished_date = workorder_best_finished_date
+            # Prepare next step
+            start_date = step_worst_finished_date
+            workorder_ids = workorder_ids.dependent_ids
         self.with_context(force_date=True).write({
-            'date_planned_start': self.workorder_ids[0].date_planned_start,
-            'date_planned_finished': self.workorder_ids[-1].date_planned_finished
+            'date_planned_start': min([workorder.leave_id.date_from for workorder in self.workorder_ids]),
+            'date_planned_finished': max([workorder.leave_id.date_to for workorder in self.workorder_ids])
         })
+        self._start_initial_workorders()
```

Comment:
the method *_plan_workorders* is still very fat with 100 lines of code. Quite bad for overrides.
Also the *_split_productions* is even worse with 300 lines of code. Methods below 50 lines of code or so would be better to allow clean overrides that are still able to call super() (otherwise they become incompatibles).